src/khepri_projection.erl

```diff
@@ -0,0 +1,296 @@
+            end,
+            FunOptions = #{ensure_instruction_is_permitted =>
+                           fun khepri_tx_adv:ensure_instruction_is_permitted/1,
+                           should_process_function => ShouldProcessFunction,
+                           is_standalone_fun_still_needed =>
+                           fun(_Params) -> true end},
+            StandaloneFun = khepri_fun:to_standalone_fun(Fun, FunOptions),
+            {extended, StandaloneFun}
+    end,
+    #?MODULE{name = Name,
+             project_fun = ProjectFun,
+             ets_options = EtsOptions}.
+
+-spec to_ets_options(Key, Value, Acc) -> Acc
+    when
+      Key :: atom(),
+      Value :: atom() | pos_integer() | boolean(),
+      Acc :: [tuple() | atom()].
+%% @hidden
+
+to_ets_options(type, Type, Acc)
+  when Type =:= set orelse Type =:= ordered_set orelse
+       Type =:= bag orelse Type =:= duplicate_bag ->
+    [Type | Acc];
+to_ets_options(keypos, Pos, Acc) when Pos >= 1 ->
+    [{keypos, Pos} | Acc];
+to_ets_options(read_concurrency, ReadConcurrency, Acc)
+  when is_boolean(ReadConcurrency) ->
+    [{read_concurrency, ReadConcurrency} | Acc];
+to_ets_options(write_concurrency, WriteConcurrency, Acc)
+  when is_boolean(WriteConcurrency) orelse WriteConcurrency =:= auto ->
+    [{write_concurrency, WriteConcurrency} | Acc];
+to_ets_options(compressed, true, Acc) ->
+    [compressed | Acc];
+to_ets_options(compressed, false, Acc) ->
+    Acc;
+to_ets_options(Key, Value, _Acc) ->
+    throw({unexpected_option, Key, Value}).
+
+-spec name(Projection) -> Name when
+      Projection :: projection(),
+      Name :: atom().
+%% @doc Returns the name of the given `Projection'.
+%% @end
+%% This is a utility that's necessary because the `khepri_projection'
+%% record is opaque.
+
+name(#?MODULE{name = Name}) ->
+    Name.
+
+-spec init(Projection) -> Ret when
+      Projection :: projection(),
+      Ret :: ok | {error, exists}.
+%% @hidden
+%% Initializes a projection. The current implementation creates an ETS
+%% table using the projection's `name/1' and {@link options()}.
+
+init(#?MODULE{name = Name, ets_options = EtsOptions}) ->
+    case ets:info(Name) of
+        undefined ->
+            _ = ets:new(Name, EtsOptions),
+            ok;
+        _Info ->
+            {error, exists}
+    end.
+
+-spec trigger(Projection, Path, Payload, Payload) -> Ret when
+      Projection :: projection(),
+      Path :: khepri_path:native_path(),
+      Payload :: term() | ?NO_PAYLOAD,
+      Ret :: ok.
+%% @hidden
+%% Applies the projection function against the entry from the tree to return
+%% a projected record. This projected record is then applied to the ETS table.
+
+trigger(
+  #?MODULE{name = Name, project_fun = ProjectFun},
+  Path, OldPayload0, NewPayload0) ->
+    Table = ets:whereis(Name),
+    OldPayload = ?PAYLOAD_OR(OldPayload0, undefined),
+    NewPayload = ?PAYLOAD_OR(NewPayload0, undefined),
```

Comment:
See my comment earlier. In the rest of the library, we use the node properties map as the structure to carry the node details around, even if it's only the payload.